pygsti/tools/optools.py

```diff
@@ -57,6 +57,65 @@ def _flat_mut_blks(i, j, block_dims):
     return ret
 
 
+def check_rank_one_density(mat, scalar_tol: Optional[Union[float,_np.floating]]=None) -> tuple[int, _np.ndarray]:
+    """
+    mat is Hermitian of order n. This function uses an O(n^2) time randomized algorithm to
+    test if mat is a PSD matrix of rank 0 or 1. It returns a tuple (r, vec), where
+
+        If r == 0, then vec is the zero vector. Either mat's numerical rank is zero
+        OR the projection of mat onto the set of PSD matrices is zero.
+
+        If r == 1, then mat is a PSD matrix of numerical rank one, and vec is mat's
+        unique nontrivial eigenvector.
+
+        If r == 2,  then our best guess is that mat's (numerical) rank is at least two
+        In exact arithmetic, this "guess" is correct with probability one. The value of
+        `vec` is ignored but provided for type-checking purposes. Additional computations
+        will be needed to determine if mat is PSD.
+
+    Conceptually, this function just takes a single step of the power iteration method
+    for estimating mat's largest eigenvalue (with size measured in absolute value).
+    See https://en.wikipedia.org/wiki/Power_iteration for more information.
+    """
+    if scalar_tol is None:
+        scalar_tol = _np.finfo(mat.dtype).eps ** 0.5
+    # ^ use for checks that have no dimensional dependence; about 1e-8 for double precision.
+    vec_tol = (mat.shape[0] ** 0.5) * scalar_tol
+    # ^ use for checks that do have dimensional dependence (will naturally increase for larger matrices)
+
+    n = mat.shape[0]
+
+    if _np.linalg.norm(mat) < vec_tol:
+        # We prefer to return the zero vector instead of None to simplify how we handle
+        # this function's output.
+        return 0, _np.zeros(n, dtype=complex)
+
+    rng = _np.random.default_rng(0)
```

Comment:
Naive question, is it fine that this has a fixed seed? I.e. that all matrices of a given dimension N use the same test vector for the power iteration.

Reply:
One could argue both sides. The worst-case scenario is that 
```python
    candidate_v = mat @ test_vec
    candidate_v /= _np.linalg.norm(candidate_v)
```
divides by zero. This would require an adversarial input.

I'll add a check to guard against this and return with a different status code if we encounter this situation. I'll also think about renaming the function to be clearer about it's role: if the returned value of ``r`` is something other than zero or one then the onus is on the calling function to do a more expensive check, like an eigendecomposition.

Reply:
renamed to fast_density_rank.